### PR TITLE
feat(aws-lambda): added support for SSM parameter store

### PR DESCRIPTION
**Pull request by kirrg001**
refs 6483

Added support for AWS SSM.

Instead of defining the plain `INSTANA_AGENT_KEY` as env variable, the customer can use the parameter store.
They can set `INSTANA_SSM_PARAM_NAME=/my/ssm/path`. We try to fetch the value from the AWS API. The challenge here is that the API call is async.

**Attention**: If you set both env variables, `INSTANA_AGENT_KEY` is always stronger. Can be discussed

### CODE REVIEW

File: packages/aws-lambda/src/ssm.js
Comment (kirrg001) [inline packages/aws-lambda/src/ssm.js:25]: Maybe consider adding a global timeout vsriablr

init vs waitAndGet -> timeout between these cllas

File: packages/aws-lambda/test/ssm_test.js
Comment (kirrg001) [inline packages/aws-lambda/test/ssm_test.js:3]: Fix dates tomorrow

Comment (willianpc) [reply packages/aws-lambda/test/ssm_test.js:3]: You can remove the Instana Copyright, I guess.
It is being removed from the lint rule.
Also, as this is a new file, you should probably change 2021 to 2022 :)

File: packages/aws-lambda/src/ssm.js
Comment (willianpc) [inline packages/aws-lambda/src/ssm.js:132]: Maybe `stopIntervalAfterMs` should have a max value, as it can be set by customers through `process.env.INSTANA_LAMBDA_SSM_TIMEOUT_IN_MS`.
If the value is too high, it could cause the setInterval ticks to run for a long period of time.
I don't have enough experience with AWS Lambda, but couldn't this hold the process for too long or cause some kind of unwanted behavior if the interval ticks continue for too long? I believe that "too long" here means 1 or 2 seconds, or something like this, no?

Comment (willianpc) [reply packages/aws-lambda/src/ssm.js:132]: Or maybe we can have a third condition that guarantees that the setInterval tick loop ends at some point.

**Review by willianpc — APPROVED** [review]
Overall, looks good to me.
I added some comments that I consider important to review, in regards of environment variables and potential setInterval long looping ticks.
Some other comments are minor suggestions that don't impact the implementation, so I leave it to you to decide if it should be changed or not.

Cheers!

File: packages/aws-lambda/src/ssm.js
Comment (kirrg001) [inline packages/aws-lambda/src/ssm.js:98]: Feedback: consistent usage of `module.exports`

File: packages/aws-lambda/src/ssm.js
Comment (basti1302) [inline packages/aws-lambda/src/ssm.js:24]: How would the env var change during the lifetime of a Lambda handler instance? I think it can only change when a new Lambda handler with a different environment configuration has been deployed, which will start a new Lambda handler instance, and thereby also restart the underlying container/process. IMO it would be valid to cache the result of this check globally.

File: packages/aws-lambda/src/ssm.js
Comment (kirrg001) [reply packages/aws-lambda/src/ssm.js:35]: Yeah consistent type returns is better.
Will change :)

File: packages/aws-lambda/src/ssm.js
Comment (basti1302) [reply packages/aws-lambda/src/ssm.js:60]: You can also combine both, like `eslint-disable-next-line $ruleId`.

Comment (kirrg001) [reply packages/aws-lambda/src/ssm.js:60]: Sure will do 👍

File: packages/aws-lambda/src/ssm.js
Comment (kirrg001) [reply packages/aws-lambda/src/ssm.js:64]: This var is a default var in AWS lambda. Customer does not need to do anything :) 
See https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html

> Lambda runtimes set several environment variables during initialization.

File: packages/aws-lambda/src/ssm.js
Comment (kirrg001) [reply packages/aws-lambda/src/ssm.js:111]: I only added `INSTANA_LAMBDA_SSM_AWS_TIMEOUT_IN_MS` for the test env to reduce the execution time.
We won't document the env variable. 

I think I'll remove this variable and switch to function options :)

File: packages/aws-lambda/src/ssm.js
Comment (kirrg001) [reply packages/aws-lambda/src/ssm.js:112]: Yeah good idea. Will add the value to the error message :)

File: packages/aws-lambda/src/ssm.js
Comment (basti1302) [reply packages/aws-lambda/src/ssm.js:64]: I think we should rather pass the region into the AWS.SSM() constructor instead of making this change globally. 

I think new `AWS.SSM({region: process.env.AWS_REGION});` should work as well and it might be less invasive.

See the section "Constructor Details" on https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SSM.html

File: packages/aws-lambda/src/ssm.js
Comment (kirrg001) [reply packages/aws-lambda/src/ssm.js:122]: Yeah same as above :) Will switch to function options
Thanks for feedback

File: packages/aws-lambda/src/ssm.js
Comment (kirrg001) [reply packages/aws-lambda/src/ssm.js:132]: > as it can be set by customers through process.env.INSTANA_LAMBDA_SSM_TIMEOUT_IN_MS.

Same as above :) Will switch to function options. Customer should not set these values

File: packages/aws-lambda/src/ssm.js
Comment (basti1302) [inline packages/aws-lambda/src/ssm.js:8]: Is there a specific need to make the parameter name configurable? Or could we also just always fetch the same parameter name and tell users which parameter they need to set?

File: packages/aws-lambda/src/ssm.js
Comment (basti1302) [inline packages/aws-lambda/src/ssm.js:70]: I am not 100% sure about that. Since the use case is storing the Instana agent key securely, we might also expect users to store them as a secure key?

File: packages/aws-lambda/src/ssm.js
Comment (kirrg001) [reply packages/aws-lambda/src/ssm.js:64]: Yeah great idea 👍

File: packages/aws-lambda/src/ssm.js
Comment (kirrg001) [reply packages/aws-lambda/src/ssm.js:8]: Good question :) 

Two reasons why I think we want to leave it as it is

1. Naming patterns. If you add a new value to the parameter store you need to choose a name.
I can imagine that companies use naming patterns such as `/company/lambda/x` 

2. Visibility. Setting INSTANA_SSM_PARAM_NAME in their env shows them that they explicitly use SSM over the default way. 

I am happy to discuss that topic on a call if you think we want to define a name for the customers :)

Comment (basti1302) [reply packages/aws-lambda/src/ssm.js:8]: OTOH, maybe letting users configure the key might be more user friendly. TBH, I am not familiar enough with SSM to know how users would expect this to work (configurable or fixed key). Feel free to ignore this comment.

File: packages/aws-lambda/src/ssm.js
Comment (kirrg001) [reply packages/aws-lambda/src/ssm.js:70]: Yeah you are right. Will add support for decrypt too :) 

From https://docs.aws.amazon.com/kms/latest/developerguide/services-parameter-store.html

> To decrypt the parameter value before returning it, set the WithDecryption parameter of GetParameter to true. When you use WithDecryption, Parameter Store calls the AWS KMS Decrypt operation on your behalf to decrypt the parameter value. As a result, the GetParameter request returns the parameter with a plaintext parameter value, as shown in the following example.

File: packages/aws-lambda/src/ssm.js
Comment (basti1302) [reply packages/aws-lambda/src/ssm.js:8]: > I am happy to discuss that topic on a call if you think we want to define a name for the customers :)

No, I think that's fine. We would still want an env var anyway to enable SSM usage (something like `INSTANA_USE_SSM=true`) and if keys can have a structure as you pointed out, letting users configure it absolutely makes sense. 👍

File: packages/aws-lambda/src/ssm.js
Comment (basti1302) [reply packages/aws-lambda/src/ssm.js:70]: Do we need both? Or only encrypted? If we offer both, we probably need a way for users to tell us which one it is, right?

File: packages/aws-lambda/src/ssm.js
Comment (kirrg001) [reply packages/aws-lambda/src/ssm.js:24]: > I think it can only change when a new Lambda handler with a different environment configuration has been deployed

Yeah I was't sure about that. Usually it works like that - a restart of a the node process is required
My thinking was that it won't hurt to re-read the value (performance wise) if we are not 100% sure

But let me just remove the two lines and cache the value :)

File: packages/aws-lambda/src/ssm.js
Comment (kirrg001) [reply packages/aws-lambda/src/ssm.js:70]: Added discussion to the ticket. Waiting for reply

Comment (kirrg001) [reply packages/aws-lambda/src/ssm.js:70]: Discussed: we want to support both

We might want to introduce an extra env variable e.g. `INSTANA_SSM_DECRYPTION=Boolean`

Any other suggestions?

**Review by basti1302 — APPROVED** [review]

**Pull request merged**
